docs/root/faq/extensions/contract.rst

```diff
@@ -0,0 +1,43 @@
+.. _faq_filter_contract:
+
+Is there a contract my HTTP filter must adhere to?
+--------------------------------------------------
+
+* Headers encoding/decoding
+
+  * During encoding/decoding of headers if a filter returns ``FilterHeadersStatus::StopIteration``,
+    the processing can be resumed if ``encodeData()``/``decodeData()`` return
+    ``FilterDataStatus::Continue`` or by explicitly calling
+    ``continueEncoding()``/``continueDecoding()``.
+
+  * During encoding/decoding of headers if a filter returns
+    ``FilterHeadersStatus::StopAllIterationAndBuffer`` or
+    ``FilterHeadersStatus::StopAllIterationAndWatermark``, the processing can be resumed by calling
+    ``continueEncoding()``/``continueDecoding()``.
+
+  * A filter's ``decodeHeaders()`` implementation must not return
+    ``FilterHeadersStatus::ContinueAndEndStream`` when called with ``end_stream`` set to *true*. In this case
+    ``FilterHeadersStatus::Continue`` should be returned.
+
+  * A filter's ``encode100ContinueHeaders()`` must return ``FilterHeadersStatus::Continue`` or
+    ``FilterHeadersStatus::StopIteration``.
+
+* Data encoding/decoding
+
+  * During encoding/decoding of data if a filter returns
+    ``FilterDataStatus::StopIterationAndBuffer``, ``FilterDataStatus::StopIterationAndWatermark``,
+    or ``FilterDataStatus::StopIterationNoBuffer``, the processing can be resumed if
+    ``encodeData()``/``decodeData()`` return ``FilterDataStatus::Continue`` or by explicitly
+    calling ``continueEncoding()``/``continueDecoding()``.
+
+* Trailers encoding/decoding
+
+  * During encoding/decoding of trailers if a filter returns ``FilterTrailersStatus::StopIteration``,
+    the processing can be resumed by explicitly calling ``continueEncoding()``/``continueDecoding()``.
+
+Are there well-known headers that will appear in the given headers map of ``decodeHeaders()``?
+----------------------------------------------------------------------------------------------
+
+Yes. The map must contain the following headers: ``Host``, ``Path`` (this might be omitted for
+CONNECT requests).
```

Comment:
@alyssawilk does this cover all relevant aspects of the contract here?

Reply:
@mattklein123 @antoniovicente as well.

Reply:
Currently the Envoy constraints are that the first filter receiving decodeHeaders will have a Host and if non-CONNECT it will have a Path.  That said there's nothing stopping a user-defined filter from removing either (though the codec may crash or throw when trying to serialize incomplete headers if a later filter doesn't add it back) as the HCM will not validate before passing to the next filter.

Reply:
Thanks for the clarification.
I've updated the text as follows:

The first filter of the decoding filter chain will have the following headers in the map:

* ``Host``
* ``Path`` (this might be omitted for CONNECT requests).

Although these headers may be omitted by one of the filters on the decoding filter chain,
they should be reinserted before the terminal filter is triggered.